# <center>Introduction to Programming with Python</center>
![Python](https://www.python.org/static/community_logos/python-logo-generic.svg)

![xkcd](https://imgs.xkcd.com/comics/python.png)

# Table of Contents <a class="anchor" id="toc">

* [Python History](#python-history)
* [What is Programming?](#what-is-programming)
* [Python Installation](#python-install)
  * [Hello World](#hello-world)
* [Jupyter Notebook Basics](#jupyter-basics)
* [Values](#values)
* [Variables](#variables)
* [Arithmetic](#arithmetic)
* [Strings](#strings)
* [Learning Resources](#learning-resources)

## Python History <a class="anchor" id="python-history">
    
[Back to Table of Contents](#toc)


### Python created by Guido von Rossum in early 1990s  (later at Google,Dropbox)

![Guido](https://upload.wikimedia.org/wikipedia/commons/thumb/d/d0/Guido-portrait-2014-curvves.jpg/290px-Guido-portrait-2014-curvves.jpg)

### Python is now programming language # 1 in TIOBE language index (as of July 2022)
https://www.tiobe.com/tiobe-index/

https://developers.slashdot.org/story/18/09/08/1722213/python-displaces-c-in-tiobe-index-top-3

## What is Programming? <a class="anchor" id="what-is-programming">
    
[Back to Table of Contents](#toc)

* Egg algorithm
* Computers are stupid, they only do what they are told to do
* If it is stupid but it works, then it is not stupid
* Make it work, make it right, make it fast (last two steps often not required in real life)
* GIGO principle

* Error messages are nothing to be afraid of, usually the message will explain what needs fixing!

## Python Installation <a class="anchor" id="python-install">
    
[Back to Table of Contents](#toc)
    
a) Python website
https://www.python.org/downloads/

b) Anaconda (includes many additional libraries)
https://www.anaconda.com/download/

c) WinPython (Windows only, some libraries included)
https://winpython.github.io/

### Language reference: https://docs.python.org/3/index.html

### "Batteries included"
![Batteries](https://github.com/ValRCS/BSSDH_22/blob/main/notebooks/img/batteries_small.jpg?raw=1)

In [1]:
# Our first comment
# Real Program Comments should generally describe why
# Here comments will describe extra information not covered or needed for starting out

# Python - Interpreted Language (commands executed as they come)

# REPL(Read,Eval,Print,Loop) - interaction with Python interpreter

## Hello World <a class="anchor" id="hello-world">
    
   [Back to Table of Contents](#toc)

In [2]:
print("Hello world!")

Hello world!


In [3]:
### Try printing a greeting of your own!
print("Hello GeoPython 2022!")

Hello GeoPython 2022!


In [6]:
### What Happens when you get an error?
print("not good")

not good


## Jupyter Notebook Basics <a class="anchor" id="jupyter-basics">
    
[Back to Table of Contents](#toc)


* Esc-M turns cell into Markdown cell for formatting (https://guides.github.com/pdfs/markdown-cheatsheet-online.pdf)
* Esc-Y turns cell into code cell(default)


* Ctrl-Enter runs code of cell in place
* Alt-Enter runs code for current cell and creates a new cell below
* Esc-A creates a new cell above current cell
* Esc-B creates a new cell below current cell
* Esc-dd deletes current cell

Note: These shortcuts work in local (Anaconda) installation of Jupyter Notebooks. Some shortcuts might not work in Google Colab.

In [ ]:
# Try Esc-B to create a new cell, 
# Enter print("Hello there!")
# Press Ctrl-Enter
# Did you get any error messages?

In [7]:
print(42)

42


In [8]:
10001

10001

In [9]:
print(10001)

10001


## Values <a class="anchor" id="values">
    
[Back to Table of Contents](#toc)

In [10]:
# Integer value (vesels skaitlis)
42

42

In [11]:
type(42)

int

In [12]:
# Floating point value (daļskaitlis)
3.14159

3.14159

In [13]:
type(3.14159)

float

In [14]:
# Text string value (teksts)
"Hello world"

'Hello world'

In [15]:
type("Hello world")

str

In [16]:
# Text values can use both single quotes ' and double quotes "
'Šī arī ir teksta rinda'

'Šī arī ir teksta rinda'

In [18]:
print("""
Šis teksts
aizņem vairākas
rindas
""")


Šis teksts
aizņem vairākas
rindas



In [19]:
# Boolean value (loģiskā vērtība) = True (patiess) or False (nepatiess)
True

True

In [20]:
type(True)

bool

In [22]:
True and False

False

## Variables <a class="anchor" id="variables">
    
[Back to Table of Contents](#toc)

In [23]:
myname = "Uldis"
# Creating our first variable will persist through this workbook once it is run

In [24]:
print(myname)

Uldis


In [25]:
y = 2019

In [26]:
theAnswer = 42

In [27]:
myPi = 3.14159

In [28]:
isHot = True

In [29]:
# type(variableName) will return variable data type
type(theAnswer)

int

In [ ]:
# What is the data type of myName ?
# How about data type of isHot ?

In [30]:
type(myPi)

float

In [31]:
type(isHot)

bool

In [32]:
isHot

True

In [34]:
# we can change the value stored in a variable
isHot = False

In [35]:
# ... it does not have to be the same type
isHot = "very hot"

In [36]:
# Variable names cannot be reserved keywords
help("keywords")


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               class               from                or
None                continue            global              pass
True                def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield
break               for                 not                 



In [ ]:
# Reserved Keywords

# False               class               from                or
# None                continue            global              pass
# True                def                 if                  raise
# and                 del                 import              return
# as                  elif                in                  try
# assert              else                is                  while
# async               except              lambda              with
# await               finally             nonlocal            yield
# break               for                 not                 
 

In [38]:
name1 = "ABC"

In [41]:
MyName = "Uldis"

In [42]:
# we can ask Python for help about its commands
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.




### Data types in Python 3.x

* Integers
  * type(42)
  * int
* Floating Point
  * type(3.14)
  * float
* Boolean 
  * type(True),type(False)
  * bool
* String(ordered, immutable char sequence)
  * type("OyCaramba")
  * str
* List 
  * type([1,2,63,"aha","youcanmixtypeinsidelist", ["even","nest"]])
  * list
* Dictionary(key:value pairs)
  * type({"foo":"bar", "favoriteday":"Friday"})
  * dict
* Tuple - ordered immutable sequence 
  * type(("sup",7,"dwarves"))
  * tuple
* Set (unordered collection of unique values)
  * type({"k","a","r","t","u","p","e","l","i","s"})
  * set

## More on variables
https://realpython.com/python-variables

## Arithmetic Operators <a class="anchor" id="arithmetic">
    
[Back to Table of Contents](#toc)
* `+ - * / `
* `**(power)` 
* `% modulus` 
* `//(integer division)`
* `() parenthesis for order`


In [47]:
(5+4)*3-(6/2)

24.0

In [44]:
5/2

2.5

In [45]:
5//2 # gives you whole - integer division


2

In [48]:
5 % 2 # this gives you reminder / technically called modulo (works the same for positives)

1

In [50]:
5%3

2

In [51]:
type(1)

int

In [52]:
type(14.0)

float

In [53]:
5**33 # 5 to 33rd power

116415321826934814453125

In [54]:
11**120 # no maximum anymore

92709068817830061978520606494193845859707401497097037749844778027824097442147966967457359038488841338006006032592594389655201

In [55]:
# Googol
10**100

10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

In [56]:
big_num = 10**100
big_num

10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

In [57]:
string_num = str(big_num) # we can convert anything to a string
string_num

'10000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'

In [58]:
string_num.count("0")

100

## More on operators
https://www.w3schools.com/python/python_operators.asp

## Strings <a class="anchor" id="strings">
    
[Back to Table of Contents](#toc)


* immutable
* Unicode support


* implement all common sequence operators
https://docs.python.org/3/library/stdtypes.html#typesseq-common

* string specific methods
https://docs.python.org/3/library/stdtypes.html#string-methods

In [59]:
name = "Valdis"
print(name)

Valdis


In [63]:
name = "Uldis 🤣"

In [64]:
print(name)

Uldis 🤣


In [65]:
# String length
len(name)

7

In [66]:
# How is this different from the string "name" ?
len("name")

4

In [67]:
# Getting Individual characters (counting from 0)
name[1]

'l'

In [68]:
# Getting last char
name[-1]

'🤣'

In [69]:
name[-3]

's'

In [96]:
# is "🤣" in our string?

print("🤣" in name)

True


In [97]:
# is "🤣" in our string?

print("Uld" in name)

True


### String Slicing

In [70]:
# Slicing syntax
# Start at 0 and go until but not including 3
name[0:3]

'Uld'

In [71]:
name[:3]

'Uld'

In [72]:
name[1:3]

'ld'

In [73]:
# last three characters
name[-3:] # so we slice from 3 characters in the end and go towards that end

's 🤣'

In [74]:
name[-3:-1] # this will give 3rd and 2nd character from the end BUT not the last one

's '

In [75]:
name[6], name[-1] # so Python offers two indexes one from the start starting at 0, and one from the end starting at -1

('🤣', '🤣')

In [76]:
name[9000]

IndexError: string index out of range

In [77]:
# slicing syntax actually has a 3rd optional modifier - step
name[0:6:2]

'Uds'

In [78]:
name[::2] # so I want to print all letters starting from start but skip every 2nd one

'Uds🤣'

In [79]:
# lets play with food!
food = "potatoes"
food[::2]

'ptte'

In [80]:
food[1::2] # we start with a 2nd letter and skip every 2nd letter from then on

'oaos'

In [81]:
food[1:6:2]

'oao'

In [82]:
# Pythonic way of reversing a string
food[::-1]


'seotatop'

In [87]:
# we can't modify a string this way
food[2] = "x"

TypeError: 'str' object does not support item assignment

In [88]:
# modifying strings
# unmutability
# food[2]="x" is not allowed
newfood = food[:2] + "x" + food[3:] # so we concatenate first 2 letters from food then add letter "x" and then rest of letters from food starting with the 4th characters
newfood

'poxatoes'

In [89]:
last_name = "Bojārs"
last_name

'Bojārs'

In [90]:
full_name = name + last_name
full_name

'Uldis 🤣Bojārs'

In [91]:
full_name = name + " " + last_name
full_name

'Uldis 🤣 Bojārs'

In [92]:
full_name = name[:-2] + " " + last_name
full_name

'Uldis Bojārs'

In [93]:
# 'some text', "some text again" work the same
# '''  For multiline ''' also """ and """
longstring = ''' This will be a very long string
    and a very long day you can write what you want
'''

In [94]:
# print out the longstring
longstring

' This will be a very long string\n    and a very long day you can write what you want\n'

In [95]:
print(longstring)

 This will be a very long string
    and a very long day you can write what you want



## "f-strings", “formatted string literals”

In some other languages also known as string interpolation

In [99]:
# Create myname and favfood variables with appropriate text
# Then run the cell below
# What would happen if you did not assign variables ?
myname

'Uldis'

In [98]:
favfood = "pizza"
favfood

'pizza'

In [100]:
print(f"My name is {myname} and my favorite food is {favfood} ")

# f strings in Python 3.6+ older formatting methods not covered in this course
# https://realpython.com/python-f-strings/


My name is Uldis and my favorite food is pizza 


In [101]:
# this is just a regular string and no variables will be replaced
print("My name is {myname} and my favorite food is {favfood} ") 

My name is {myname} and my favorite food is {favfood} 


In [ ]:
# Old string concatation method
print("My name is " + myname + " and my favorite food is " + favfood)

In [102]:
food_string = f"My name is {myname} and my favorite food is {favfood} "  # anything inside " is significant"
print(food_string)

My name is Uldis and my favorite food is pizza 


# Python Resources <a class="anchor" id="learning-resources">
    
[Back to Table of Contents](#toc)

## Wiki for Tutorials

https://wiki.python.org/moin/BeginnersGuide/NonProgrammers

## Tutorials Begginner to Intermediate




* https://automatetheboringstuff.com/ - Anything by Al Sweigart is great
* http://newcoder.io/tutorials/ - 5 sets of practical tutorials
* [Think Like a Computer Scientist](http://interactivepython.org/runestone/static/thinkcspy/index.html) full tutorial
* [Non-Programmers Tutorial for Python 3](https://en.wikibooks.org/wiki/Non-Programmer%27s_Tutorial_for_Python_3) quite good for wikibooks
* [Real Python](https://realpython.com/) Python Tutorials for all levels


* [Learn Python 3 the Hard Way](https://learnpythonthehardway.org/python3/intro.html) controversial author but very exhaustive, some like this approach

## More Advanced Python Specific Books

* [Python Cookbook](https://www.amazon.com/Python-Cookbook-Third-David-Beazley/dp/1449340377) Recipes for specific situations

* [Effective Python](https://effectivepython.com/) best practices
* [Fluent Python](http://shop.oreilly.com/product/0636920032519.do) **highly recommended**, shows Python's advantages

## General Best Practices Books
#### (not Python specific)

* [Code Complete 2](https://www.goodreads.com/book/show/4845.Code_Complete) - Fantastic best practices
* [The Mythical Man-Month](https://en.wikipedia.org/wiki/The_Mythical_Man-Month) - No silver bullet even after 40 years.
* [The Pragmatic Programmer](https://www.amazon.com/Pragmatic-Programmer-Journeyman-Master/dp/020161622X) - More practical advice
* [Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882) - more towards agile

## Blogs / Personalities / forums

* [Dan Bader](https://dbader.org/)
* [Reddit Python](https://www.reddit.com/r/python)

## Exercises/Challenges
* http://www.pythonchallenge.com/ - first one is easy but after that...
* [Advent of Code](https://adventofcode.com/) - yearly programming challenges
* https://projecteuler.net/ - gets very mathematical but first problems are great for testing 

## Explore Public Notebooks on Github
 Download them and try them out for yourself
 
https://github.com/jupyter/jupyter/wiki/A-gallery-of-interesting-Jupyter-Notebooks

## Questions / Suggestions ? 

This notebook is based on the Python introduction notebook by Valdis Saulespurēns:
* https://github.com/ValRCS/BSSDH_22/blob/main/notebooks/Python%20Introduction.ipynb

Pull requests welcome

e-mail **uldis.bojars at gmail.com**